### PR TITLE
 [NDSL] Aer Activation #942 

### PR DESCRIPTION
**Pull request by mattweil11**
Port of Aer Activation component of Moist to pyMoist

Original code: [aer_actv_single_moment.F90:AerActivation](https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/db55c301840d98b788b0e17045510af726c0f555/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/aer_actv_single_moment.F90#L39)

Parent: https://github.com/GEOS-ESM/SMT-Nebulae/issues/57

See state of validation at: https://github.com/GEOS-ESM/GEOSgcm_GridComp/issues/942

### CODE REVIEW

**Comment by FlorianDeconinck** [comment]
`0 diff trivial`: DSL port work, no changes to current production code

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+from gt4py.cartesian.gtscript import PARALLEL, computation, exp, interval, log, sqrt
+
+import pyMoist.aer_activation_constants as constants
+from ndsl import QuantityFactory, StencilFactory
+from ndsl.dsl.typing import Float, FloatField, FloatFieldIJ, Int
+from pyMoist.types import FloatField_NModes
+
+
+@gtscript.function
+def GammLn(xx: Float) -> Float:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:11]: Let's move this guy in a file `numerical_recipes`

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+    ser += -1.231739572450155 / (y + 1)
+    ser += 0.001208650973866179 / (y + 1)
+    ser += -0.000005395239384953 / (y + 1)
+
+    gammln = tmp + log(stp * ser / x)
+    return gammln
+
+
+@gtscript.function
+def _gser_stencil(a: Float, x: Float, gln: Float) -> Float:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:44]: Same as above. Maybe rename or is `gser` expected?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+
+    Returns:
+    Float: The series representation of the incomplete gamma function.
+    """
+    eps = 3.0e-9  # was eps=3.0d-07 in press et al.
+    itmax = 10000  # was itmax=100   in press et al.
+    gln = GammLn(a)
+    if x <= 0:
+        # Fortran messages here x < 0 in gser
+        # TODO; Allow print in GT4Py
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:63]: `TODO:`

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+            if abs(del_) < abs(sum_) * eps:
+                gamser = sum_ * exp(-x + a * log(x) - gln)
+                n = itmax
+            n += 1
+        gamser = sum_ * exp(-x + a * log(x) - gln)
+    return gamser
+
+
+@gtscript.function
+def _gcf_matrix_stencil(a: Float, x: Float, gln: Float) -> Float:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:85]: Same as above

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+        del_ = d * c
+        h *= del_
+        if abs(del_ - 1.0) < eps:
+            i = itmax + 1
+        i += 1
+    return exp(-x + a * log(x) - gln) * h
+
+
+@gtscript.function
+def GammP(a: Float, x: Float) -> Float:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:128]: Same as above

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+    gln = GammLn(a)
+    if x < a + 1.0:
+        gammp = _gser_stencil(a, x, gln)
+    else:
+        gammp = 1.0 - _gcf_matrix_stencil(a, x, gln)
+    return gammp
+
+
+@gtscript.function
+def Erf(x: Float) -> Float:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:154]: Same as above

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+    Compute the error function for a given value x.
+
+    Parameters:
+    x (Float): Input value for which the error function is to be computed.
+
+    Returns:
+    Float: The error function value for the input x.
+    """
+    erf = 0.0
+    if x < 0.0e00:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:167]: `<0.0e00` isn't that `0.0` ?

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+
+        if constants.n_modes != n_modes:
+            raise NotImplementedError(
+                f"Coding limitation: 14 modes are expected, getting {n_modes}"
+            )
+
+        if not USE_AERSOL_NN:
+            raise NotImplementedError("Non NN Aerosol not implemented")
+
+        self.higher_dimensional_storages = stencil_factory.from_origin_domain(
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:528]: rename that stencil

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/__init__.py [hunk 9]
```diff
@@ -1 +1 @@
-from .translate_radiation_coupling import TranslateRadCouple
+from .translate_aer_activation import TranslateAerActivation
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/__init__.py:1]: This is wrong. `TranslateAerActivation` should be added nothing removed

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/translate_radiation_coupling.py
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/translate_radiation_coupling.py:1]: Do not remove

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:1]: Do not remove

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py:1]: Do not remove

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh:1]: Do not replace. Make it additive to bring both data packet in a common `moist` subdirectory

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/run_tests.sh
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/run_tests.sh:1]: Do not replace. Make it additive to have both test runnable.

**Review by FlorianDeconinck — CHANGES_REQUESTED** [review]
Careful about merging code that is destructive to previous work!

Please do a pass on the documentation for the types. Instead of just `xx (Float)` also indicated `In`/`Out`/`Inout`.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py (hunk 2 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:44]: gser is name of function in Fortran so I just kept it. I removed the _stencil portion as it is not a stencil.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py (hunk 3 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:63]: Fixed, same with the TODO: in GammP

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py (hunk 4 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:85]: Got rid of the _stencil and renamed to gcf_matrix

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py (hunk 5 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:128]: Added to numerical recipes along with the other functions that contained "numerical recipes" in the docstring

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/translate_radiation_coupling.py
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/translate_radiation_coupling.py:1]: Added back

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:1]: Added back

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py:1]: Added back

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh:1]: Put everything back into pyMoist/tests_data/geos_11.5.2/moist, did not get data from radiation coupling on NCCS as port is for aer_activation

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/run_tests.sh
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/run_tests.sh:1]: Added back; user can get data for radiation coupling and change which_modules if desired

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py (hunk 1 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/aer_activation.py:11]: Put into numerical_recipes

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/__init__.py (hunk 9 above)
Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/savepoint/__init__.py:2]: TranslateRadCouple line added

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh [hunk 10]
```diff
@@ -4,8 +4,8 @@ set -e -x
...
 mkdir -p $TEST_DATA_PATH
-wget https://portal.nccs.nasa.gov/datashare/astg/smt/geos-fp/translate/11.5.2/Moist/rad_coup_data_nc.tar.gz
-mv rad_coup_data_nc.tar.gz $TEST_DATA_PATH
+wget https://portal.nccs.nasa.gov/datashare/astg/smt/geos-fp/translate/11.5.2/Moist/AerActivation_nc.tar.gz
+mv AerActivation_nc.tar.gz $TEST_DATA_PATH
 cd $TEST_DATA_PATH
-tar -xzvf rad_coup_data_nc.tar.gz
-rm rad_coup_data_nc.tar.gz
+tar -xzvf AerActivation_nc.tar.gz
+rm AerActivation_nc.tar.gz
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh:11]: Have the `get_data.sh` bring both data set into a single `moist` directory

**Review by FlorianDeconinck — CHANGES_REQUESTED** [review]
Clean up data wget so we can do AerActivation and RadCoupling at the same time

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh (hunk 10 above)
Comment (FlorianDeconinck) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/tests/scripts/get_data.sh:11]: Also look at renaming your `higher_dimensional_storage` stencil to something closer to `aerosol_activation`

**Review by FlorianDeconinck — APPROVED** [review]

**Pull request merged**